Update complete application tutorial (#569)
Avaiga/taipy-doc: docs/tutorials/complete_applications/complete_application/step_04/step_04.md

> The full code is available
<a href="./../src/src.zip" download>here</a>.

# Step 4: Scenario Page

> The Scenario Page is a section of the application made for crafting and tailoring 
> prediction scenarios using time series data. Users can adjust various parameters for the 
> prediction, like the prediction date, maximum capacity, and the quantity of predictions. 
> This page also features a chart that shows past data and predictions created using both machine learning and baseline techniques.

Reviewer: ```suggestion
The Scenario Page is a part of the application designed for creating and 
customizing prediction scenarios using time series data. Users can modify different parameters 
for the prediction, such as the prediction date, maximum capacity, 
and the number of predictions. This page also includes a chart displaying historical data 
and predictions generated using both machine learning and baseline methods.
```

    state_day = dt.datetime(state.day.year, state.day.month, state.day.day)

    # Change the default parameters by writing in the Data Nodes
    state.scenario.day.write(state_day)
    state.scenario.n_predictions.write(int(state.n_predictions))
    state.scenario.max_capacity.write(int(state.max_capacity))
    notify(state, "success", "Saved!")
    

def on_change(state, var_name, var_value):
    if var_name == "scenario" and var_value:
        state.day = state.scenario.day.read()
        state.n_predictions = state.scenario.n_predictions.read()
        state.max_capacity = state.scenario.max_capacity.read()
        
        if state.scenario.full_predictions.is_ready_for_reading:
            state.predictions_dataset = state.scenario.full_predictions.read()
        else:
            state.predictions_dataset = predictions_dataset

scenario_page = Markdown("pages/scenario/scenario.md")
```

It includes the following components:

- **Global Variables**:

The global variables *scenario*, *day*, *n_predictions*, *max_capacity*, and *predictions_dataset* are defined. These variables store the initial state of the application.

- **Save Function**:

> The `save` function is responsible for saving the current scenario state. When the user clicks the "Save" button, this function is called. It takes the state of the page as input, converts the date format to the appropriate format, and updates the scenario parameters accordingly. It then notifies the user with a success message.

Reviewer: ```suggestion
The `save` function is in charge of preserving the current scenario state. 
When the user clicks the "Save" button, this function gets activated. 
It receives the page's state as input, converts the date format to the correct one, 
adjusts the scenario parameters accordingly, and then informs the user with a success message.
```

- **On Change Function**:

> The `on_change` function is called when any variable on the page changes its value. It monitors the changes in the scenario variable and updates the other variables accordingly. It also checks if the *full_predictions* are ready for reading and updates the *predictions_dataset*` accordingly.

Reviewer: ```suggestion
The `on_change` function is called whenever any variable on the page undergoes a value change. 
It keeps track of alterations in the scenario variable and adjusts the other variables accordingly. 
It also verifies if the *full_predictions* are available for reading and updates the *predictions_dataset* accordingly.
```

> The full code is available
<a href="./../src/src.zip" download>here</a>.

The Scenario Page is a part of the application designed for creating and 

> customizing prediction scenarios using time series data. Users can modify different parameters 

Reviewer 2: ```suggestion
customizing prediction scenarios using time series data. End-users can modify different parameters 
```

for the prediction, such as the prediction date, maximum capacity, 
and the number of predictions. This page also includes a chart displaying historical data 
and predictions generated using both machine learning and baseline methods.

> ![Scenario Page](result.png){ width=700 style="margin:auto;display:block;border: 4px solid rgb(210,210,210);border-radius:7px" }"

Reviewer 2: ```suggestion
![Scenario Page](result.png){ width=700 style="margin:auto;display:block;border: 4px solid rgb(210,210,210);border-radius:7px" }
```

> The Scenario Page is constructed using a combination of Markdown and Python code. Below is a detailed explanation of each component:

Reviewer 2: ```suggestion
The Scenario Page is constructed using a combination of Markdown and Python code. Below is a detailed explanation.
```

> # Markdown (pages/scenario/scenario.md)

Reviewer 2: ```suggestion
# Markdown

The following Markdown corresponds to the `pages/scenario/scenario.md` file.
```

```markdown
# Create your scenario:

<|layout|columns=3 1 1 1 1|
<|{scenario}|scenario_selector|>

**Prediction date** <br/>
<|{day}|date|active={scenario}|not with_time|>

**Max capacity** <br/>
<|{max_capacity}|number|active={scenario}|>

**Number of predictions** <br/>
<|{n_predictions}|number|active={scenario}|>

<br/> <|Save|button|on_action=save|active={scenario}|>
|>
 
<|{scenario}|scenario|on_submission_change=submission_change|>

<|{predictions_dataset}|chart|x=Date|y[1]=Historical values|type[1]=bar|y[2]=Predicted values ML|y[3]=Predicted values Baseline|>

# Data Node Exploration

<|layout|columns=1 5|
<|{data_node}|data_node_selector|>

<|{data_node}|data_node|>
|>
```

The Markdown section outlines the arrangement and elements of the Scenario Page. It consists of the following components:

- **Scenario Selector**: `<|{scenario}|scenario_selector|>`

> A component (a dropdown) that allows users to select different scenarios. The selected scenario will affect the values of other components on the page.

Reviewer 2: Should we say something like this?
The scenario value is used in other elemtns so selecting a scenario affects the other elements.

Author: I added your sentence, this is a good idea

- **Max Capacity**: `<|{max_capacity}|number|...|>`

A number input field where users can set the maximum capacity value. This value is used to cap the predictions if they exceed the specified maximum.

- **Number of Predictions**: `<|{n_predictions}|number|...|>`

A number input field where users can set the desired number of predictions to be made.

- **Save Button**: `<|Save|button|on_action=save|active={scenario}|>`

A button that triggers the "save" action when clicked. It is used to save the selected scenario and parameter values.

- **Scenario Section**: `<|{scenario}|scenario|on_submission_change=submission_change|>`

A section that displays information about the currently selected scenario. It includes details about the scenario, properties, and the ability to delete or submit the scenario. When the scenario status is changed, `submission_change` is called with information on the scenario.

- **Predictions Chart**: `<|{predictions_dataset}|chart|...|>`

A chart that displays historical values and the predicted values obtained from machine learning and baseline methods. It shows how well the predictions align with the historical data.

> - **Data Node Exploration**: `<|{data_node}|data_node_selector|> <|{data_node}|data_node_selector|>`

Reviewer 2: Duplication

Author: It should be data_node for the second one

Reviewer 2: AH yes, indeed.

This is where the detailed information and history about the selected data node is presented. Depending on the nature of the data node, this could display raw data in a tabular format, visualizations, texts, or dates. If the format allows it, the user can directly write new values in the data node.

> # Python Code (pages/scenario/scenario.py)

Reviewer 2: ```suggestion
# Python Code 

The following Python code corresponds to the `pages/scenario/scenario.py` file.
```

    state_day = dt.datetime(state.day.year, state.day.month, state.day.day)

    # Change the default parameters by writing in the Data Nodes
    state.scenario.day.write(state_day)
    state.scenario.n_predictions.write(int(state.n_predictions))
    state.scenario.max_capacity.write(int(state.max_capacity))
    notify(state, "success", "Saved!")
    

def on_change(state, var_name, var_value):
    if var_name == "scenario" and var_value:
        state.day = state.scenario.day.read()
        state.n_predictions = state.scenario.n_predictions.read()
        state.max_capacity = state.scenario.max_capacity.read()
        
        if state.scenario.full_predictions.is_ready_for_reading:
            state.predictions_dataset = state.scenario.full_predictions.read()
        else:
            state.predictions_dataset = predictions_dataset

scenario_page = Markdown("pages/scenario/scenario.md")
```

It includes the following components:

- **Global Variables**:

The global variables *scenario*, *data_node*, *day*, *n_predictions*, *max_capacity*, and *predictions_dataset* are defined. These variables store the initial state of the application.

- **Save Function**:

> The `save` function is in charge of preserving the current scenario state. 

Reviewer 2: Should we use the term callback instead of function? 
```suggestion
The `save` function is used as a callback. It is in charge of preserving the current scenario state. 
```
Then once callback is defined, we can use callback in place of function. What do you think?
